File: w1/d4/armstrong-numbers/js/armstrongNumbers.js
// How can you make this more scalable and reusable later?

exports.findArmstrongNumbers = function (array) {
  //init answer array and length of number
  let ans = [];
  let len;
  // loop through input array
  array.forEach((num) => {
    // work through each number to see if it is an armstrong number
    num = num.toString();
    len = num.length;
    let sum = 0;
    for (a of num) {
      sum += a ** len;
    }
    if (sum === +num) {
      ans.push(+num);
    }
  });
  // return answer array
  return ans;
};

File: w1/d4/armstrong-numbers/python/armstrong_numbers.py
# How can you make this more scalable and reusable later?
import functools


def find_armstrong_numbers(numbers):
    # ans = []
    # for number in numbers:
    #     number = str(number)
    #     sum = 0
    #     length = len(number)
    #     for a in number:
    #         sum += int(a)**length
    #     if sum == int(number):
    #         ans.append(sum)

    # return ans
    # numbers = [str(x) for x in numbers]
    ans = []
    for x in numbers:
        if helper(x) == x:
            ans.append(x)
    return ans


def helper(num):
    digits = [int(x) for x in str(num)]
    length = len(digits)
    sum = functools.reduce(lambda agg, d: agg + (d**length), digits, 0)
    return sum


# helper(371)
# helper(5)